### ISSUE TITLE
[BUG] OpenBullet2.ApplicationDbContext

### ISSUE THREAD

**Issue by LazarAlts** [openbullet/OpenBullet2#235]
![image](https://user-images.githubusercontent.com/63159594/115374488-2f5e2a00-a1cd-11eb-9225-63799cc76881.png)

OS: Linux Ubuntu 18.04
Browser: Brave
Version: Latest and 0.1.10

**Comment by RydjGit** [comment]
@openbullet  I think this error is very recurrent, with all the fixes that you made.
One solution and I think it is a clean one, is to implement a singleton class that wraps access to the Database using a single dedicated `background Thread`, this thread is initialized and started at the constructor of this class with a `BlockingCollection`  that accepts a struct whose properties are (the Object to be added to database and a delegate for the add procedure)
this should increase performance significantly because ,(correct me if I'm wrong), each task you start in the parallelizer, actually when it ends, tries to access the database to insert the current position of job's wordlist, and if it cannot acquire the lock object it has to wait/sleep 10 milliseconds ,this is bad for over all performance.

  I think an ideal class for this purpose would be static , but as you know I'm not too familiar with Blazor API's , but I think there is an option to inject a singleton  class  in the startup method.

 This is a classic Multi producer single consumer scenario.
I would have written this my self but I fear breaking the logic of the overall app as Openbullet2 is a large project and I cannot keep-up with all Methods and functions.

**Comment by openbullet** [comment]
Hey, I don't think that's a good solution, or at least that's not something that is done in the enterprise-grade applications I see around. It would be a workaround, sure, but it's not the proper way to do things.

> a singleton class that wraps access to the Database using a single dedicated `background Thread`

This is not good. Entity Framework Core already takes care of the database access. An instance of `ApplicationDbContext` lives inside the repository that is passed to a job when it's created. So every job has its own separate access that takes care of storing all hits of that specific job as they come.

> each task you start in the parallelizer, actually when it ends, tries to access the database to insert the current position of job's wordlist, and if it cannot acquire the lock object it has to wait/sleep 10 milliseconds ,this is bad for over all performance.

https://github.com/openbullet/OpenBullet2/blob/master/RuriLib/Models/Jobs/MultiRunJob.cs#L598
The method that saves a hit to the database is fire and forget, the call is not awaited so there's no performance degradation due to this.

> I think there is an option to inject a singleton class in the startup method

This is bad because it would keep a long-lasting connection to the database for the entire lifetime of an application, even if nothing is accessing the database, which is bad. Repositories should be transient, not singletons, that's what everyone tells you to do and that's what I followed.

> This is a classic Multi producer single consumer scenario.

The original idea was to have 1 service bus (e.g. RabbitMQ) and then a separate process that takes care about storing the hits from the queues to the database, but this would have been hard to deploy for the average user, as it's a microservice architecture.

> I would have written this my self

Don't worry, I will take care of this. The current implementation is correct, I just need to debug the cause or put additional checks to prevent the exceptions from happening.

**Comment by openbullet** [comment]
Currently I don't have much time, see reasons here
https://discourse.openbullet.dev/t/slow-development/2190

When I'm graduated I will have more time and I will get OB2 ready to exit beta stage and move to release stage by debugging it properly, adding more logging etc.

**Comment by RydjGit** [comment]
Thank you for the great explanation,  

> An instance of ApplicationDbContext lives inside the repository that is passed to a job when it's created. So every job has its own separate access that takes care of storing all hits of that specific job as they come.

The same instance or a new dedicated instance for each job?

> The original idea was to have 1 service bus (e.g. RabbitMQ)

This a great idea indeed and has the same concept as I'm proposing.

> I just need to debug the cause or put additional checks to prevent the exceptions from happening.

That's my point exactly, performance degradation is due to thread locks, excessive checks, exceptions, Task objects.
when I work with low level API's our goal is always to keep thread locks, exceptions and if statements to a minimum as these affect performance, as for Tasks, we avoid those as well  because each Task is an object which needs GC collection. and too much GC is a big taboo for performance in microservices.

also I really have an issue with 1000 threads you "allow" the application to start .
each threads consumes 1 MB of memory and with 1000 background threads and 1000 IO threads that's 2 GB of memory wasted because no way the application needs all these threads and in my Testing removing the statements I saw some memory consumption improvements.

I like this Talk by Jeffrey Richter please watch it, if you have some spare time, it is very informative:

[https://www.youtube.com/watch?v=hB0K1JWFoqs](url)

Finally I wish you the best with your studies and don't prioritize anything before getting your degree.

**Comment by openbullet** [comment]
> The same instance or a new dedicated instance for each job?

Actually, now that you make me think of it, I see there's something terribly wrong in the logic.

This class https://github.com/openbullet/OpenBullet2/blob/master/OpenBullet2/Services/JobFactoryService.cs is instanced as a singleton in `Startup.cs`, which is wrong. It's passing the same `IHitRepository` to new instances of `HitOutputFactory`, each with its separate semaphore, but given out to different jobs. So if there are 2 jobs that write hits to the database there will be exceptions. Now I see why the exceptions are happening, thank you for making me look into this. Maybe the service bus approach won't be bad after all, although it will have to be an in-memory service bus since I cannot expect users to deploy RabbitMQ by themselves.

> also I really have an issue with 1000 threads you "allow" the application to start

I tried not setting that value, but it would just check really slowly. That's why I added it in the first place, because I saw a huge CPM increase. But I agree that it's a very dirty trick.

> I like this Talk by Jeffrey Richter please watch it, if you have some spare time, it is very informative:

Thank you for the suggestion, I will definitely watch it!

> Finally I wish you the best with your studies and don't prioritize anything before getting your degree.

Thanks for understanding ^_^ When I have more time I would really like to have a chat with you to discuss about OB2 and what else can be improved, since you seem to be a very experienced developer with an eye for performance and low level code.
